### DISCUSSION TITLE
ORA-08192: Flashback Table operation is not allowed on fixed tables

### DISCUSSION THREAD

**Discussion by JesperSiteimprove** [godror/godror#187] [Q&A]
Hi

I am facing a problem using godror where I fetch some dates and try to insert them in a table.

The date: to_date('01-01-0001','dd-mm-yyyy')
Seems to be causing the problem.

I only get the error when using 
github.com/godror/godror v0.24.1
or higher.

github.com/godror/godror v0.24.0
or lower do not show the error.

The following test triggers the ora-08192

```
package main

import (
	"context"
	"database/sql"
	"os"
	"testing"
	"time"

	"github.com/godror/godror"
)

func TestForError8192(t *testing.T) {
	dbConnectString := os.Getenv("ORA_CONN_STRING")

	params, err := godror.ParseConnString(dbConnectString)
	if err != nil {
		t.Fatal(err)
	}
	params.StandaloneConnection = true
	db, err := sql.Open("godror", params.StringWithPassword())
	if err != nil {
		t.Fatal(err)
	}

	ctx, cancel := context.WithTimeout(context.Background(), 120*time.Second)
	defer cancel()

	conn, err := db.Conn(ctx)
	if err != nil {
		t.Fatal(err)
	}

	// prepare table
	if _, err := conn.ExecContext(ctx, "DROP TABLE test_date_error"); err != nil {

	}
	if _, err := conn.ExecContext(ctx, "CREATE TABLE test_date_error (problem_ts date)"); err != nil {
		t.Fatal(err)
	}

	// fetch a date from the database
	selectStmt, err := conn.PrepareContext(ctx, "select to_date('01-01-0001','dd-mm-yyyy') problem_ts from dual")
	if err != nil {
		t.Fatal(err)
	}
	defer selectStmt.Close()
	rows, err := selectStmt.Query()
	if err != nil {
		t.Fatalf("Failed to query: %s\n", err)
	}
	defer rows.Close()

	var problemT time.Time
	for rows.Next() {
		err = rows.Scan(&problemT)
	}
	err = rows.Err()

	// insert the date in the database
	tx, err := conn.BeginTx(ctx, nil)
	if err != nil {
		t.Fatalf("Failed to beginTx: %s\n", err)
	}
	defer tx.Rollback()

	stmt, err := tx.PrepareContext(ctx, "insert into test_date_error(problem_ts) values(:problem_ts)")
	if err != nil {
		t.Errorf("Failed to prepare insSQL %s", err)
		_ = stmt.Close()

	}
	defer stmt.Close()

	_, err = stmt.Exec(sql.Named("problem_ts", problemT))
	if err != nil {
		t.Fatalf("exec failure: %v\n", err)
	}

	tx.Commit()
}

```
run it like:
`ORA_CONN_STRING=schema/pw@sid go test -v -run TestForError8192`

Hope you can help
Best Regards
Jesper Persson

**Comment by tgulacsi** [comment]
What is the error, end where does it appear?
TestForError8192 in c4deb63 does not produce it.

**Reply by JesperSiteimprove** [reply]
The error happens in the "exec failure" line and the error is: `ORA-08192: Flashback Table operation is not allowed on fixed tables`

I can see that I also do not get the error running the TestForError8192 in [c4deb63](https://github.com/godror/godror/commit/c4deb63640d32e50186ab33857d40e7c7266c8c2)

After comparing your connectionstring with the one I am using, I could make the error go away if I set the `timezone` parameter. So the error only happens if the timezone parameter is not set in the connectionstring.

**Reply by tgulacsi** [reply]
`to_date('01-01-0001','dd-mm-yyyy')` is very close to `time.Time{}`, and the zero time (`time.IsZero()` - https://play.golang.org/p/aG9lRwY6-qW) is treated as NULL - maybe this causes your error.

Why do you need that specific date? Adding a second may help.
Or use sql.NullTime, which does not have this problem.

**Reply by JesperSiteimprove** [reply]
That date has special meaning in some of our tables and columns as the default, so it is not easy just to change that part. 
Using `var problemT sql.NullTime` by it self does not make the error go away.
But when combined with setting the `timezone=UTC` on the `params` then I dont see the error and the correct date is inserted in the `test_date_error(problem_ts)` column.
I am not sure exactly what side-effect it can have to just always use `timezone=UTC`, as I am trying now, but I haven't seen any unwanted behaviour in my program.

**Reply by ykh-siteimprove** [reply]
Oracle support answer: Can you check with your application if "`OCIDATETIMECONSTRUCT`" is used and if an invalid date/time/month is passed only then issue is reproduced?  `OCIDateTimeConstruct` is a code piece that constructs a date and calls another function to check its validity. You may check with the application if they use this `OCIDateTimeConstruct` API and pass and may be wrongly passing an invalid month/date/

**Reply by tgulacsi** [reply]
AFAIS godror does not use OCIDateTimeConstruct (dpiOci___dateTimeConstruct, dpiDataBuffer__toOracleTimestamp) directly, but uses dpiVar__setValue, which calls dpiDataBuffer__toOracleTimestamp, which calls dpiOci___dateTimeConstruct, if the value to be set is a DPI_NATIVE_TYPE_TIMESTAMP, which is used for time.Time.

What is your database's time zone? What is your local timezone?

Setting timezone in connection param sets the database's time zone from the perspective of the driver. It affects converting DATE types to time.Time, treating the DATE in the DB's time zone when retrieving from the database.

**Reply by ykh-siteimprove** [reply]
> What is your database's time zone?

In the DB OS:
```
$ timedatectl | grep 'Time zone:'
       Time zone: Europe/Copenhagen (CEST, +0200)

```

In the DB:
```
SQL> select DBTIMEZONE from dual;

DBTIME
------
+02:00

```

This function returns a **static** result - the time offset from `UTC` at the time of the `CREATE DATABASE ...` command.
And this does not reflect future daylight savings time.


> What is your local timezone?

On the client?
I believe Jesper is using `Europe/Copenhagen`.


> Setting timezone in connection param sets the database's time zone from the perspective of the driver.

We are not using such settings. At least explicitly.

**Reply by sudarshan12s** [reply]
> AFAIS godror does not use OCIDateTimeConstruct (dpiOci___dateTimeConstruct, dpiDataBuffer__toOracleTimestamp) directly, but uses dpiVar__setValue, which calls dpiDataBuffer__toOracleTimestamp, which calls dpiOci___dateTimeConstruct, if the value to be set is a DPI_NATIVE_TYPE_TIMESTAMP, which is used for time.Time.
> 
> What is your database's time zone? What is your local timezone?
> 
> Setting timezone in connection param sets the database's time zone from the perspective of the driver. It affects converting DATE types to time.Time, treating the DATE in the DB's time zone when retrieving from the database.

just to confirm,  yes dpiOci__dateTimeConstruct calls OCIDateTimeConstruct. In case of error ORA-01841  which i get on my client mc, I see timezone=-07:00 ,year=0 . For working case (timezone=UTC in connectstring) year is not 0 . year=1, timezone=+00:00. passing year as 0 was causing error ORA-01841.

I see following as timezone information in Go connection (c.params.Timezone) which is used to pass to C.dpiData_setTimestamp. year is 0 passed to dpiData_setTimestamp.  c.Timezone() is 0000-12-31 17:00:00 -0700 -07:00

**Reply by sudarshan12s** [reply]
some reason i cant reproduce same error ORA-8192. 
I get this error, ORA-01841: (full) year must be between -4713 and +9999, and not be 0 . 
where stmt.go passes Year as 0 for C.dpiData_setTimestamp function and goes away once I do this change in z_test.go

```
  //_, err = stmt.Exec(sql.Named("problem_ts", sql.NullTime{Valid: true}))
    _, err = stmt.Exec(sql.Named("problem_ts", problemT))
```

@JesperSiteimprove, Can you add some prints  to see what values are  passed to C.dpiData_setTimestamp in stmt.go@1411 line

        + fmt.Println("Time user passed", t)
        t = t.In(c.Timezone())
        + fmt.Println(" Timezone used in connection ", c.Timezone())
        + fmt.Println(" Time passed to dpiData_setTimestamp ", t)

**Reply by JesperSiteimprove** [reply]
@sudarshan12s 
It seems I can only get the error against our own test database.

---------------------------------------------------------
-- FAILS when not setting timezone in connection

```
GODROR_TEST_DSN="user=xxx password=xxx connectString=dkdt1t"  go test -v -run TestForError8192
Version: v0.26.3
Client: 12.2.0.1.0 Timezone: Local
Server: 19.0.0.0.0 [Oracle Database 19c Standard Edition 2 Release 19.0.0.0.0 - Production] Timezone: Local
=== RUN   TestForError8192
Time user passed 0001-01-01 00:00:00 +0050 LMT
 Timezone used in connection  Local
 Time passed to dpiData_setTimestamp  0001-01-01 00:00:00 +0050 LMT
    z_test.go:4198: exec failure: ORA-08192: Flashback-operation på tabel er ikke tilladt på faste tabeller
--- FAIL: TestForError8192 (0.88s)
FAIL
exit status 1
FAIL	github.com/godror/godror	7.132s
```

---------------------------------------------------------
-- WORKS when setting timezone in connection

```
GODROR_TEST_DSN="user=xxx password=xxx connectString=dkdt1t timezone=UTC"  go test -v -run TestForError8192
Version: v0.26.3
Client: 12.2.0.1.0 Timezone: Local
Server: 19.0.0.0.0 [Oracle Database 19c Standard Edition 2 Release 19.0.0.0.0 - Production] Timezone: Local
=== RUN   TestForError8192
Time user passed 0001-01-01 00:00:00 +0000 UTC
 Timezone used in connection  UTC
 Time passed to dpiData_setTimestamp  0001-01-01 00:00:00 +0000 UTC
--- PASS: TestForError8192 (0.77s)
PASS
ok  	github.com/godror/godror	3.582s
```

---------------------------------------------------------
My local setup:

```
$ cat /etc/timezone
Europe/Copenhagen

$ date +%Z
CEST

$ date +"%Z %z"
CEST +0200

$ timedatectl
               Local time: Mon 2021-10-11 06:57:42 CEST   
           Universal time: Mon 2021-10-11 04:57:42 UTC    
                 RTC time: Mon 2021-10-11 04:57:42        
                Time zone: Europe/Copenhagen (CEST, +0200)
System clock synchronized: yes                            
              NTP service: n/a                            
          RTC in local TZ: no 
```

---------------------------------------------------------
-- our test database setup

```
> select DBTIMEZONE from dual;

DBTIME
------
+01:00

$ timedatectl | grep 'Time zone:'
       Time zone: Europe/Copenhagen (CEST, +0200)

```

**Reply by tgulacsi** [reply]
Where's that LMT +5 coming from?
What if you try `time.Time{}.Add(-5*time.Hour).In(time.FixedZone("LMT", +5*3600))` with setting the timezone? (this is the "0001-01-01 00:00:00 +0050 LMT")
AFAIK your correct "timezone" setting would be +0200 BTW.

**Reply by JesperSiteimprove** [reply]
I must admit I dont know where the `LMT +0050` is coming from, same reason I send you my local setup of timezone

If I insert this before the sql.Open:
`params.Timezone = time.Time{}.Add(-5 * time.Hour).In(time.FixedZone("LMT", +5*3600)).Location()`

Then I get the error:
```
Version: v0.26.3
Client: 12.2.0.1.0 Timezone: Local
Server: 19.0.0.0.0 [Oracle Database 19c Standard Edition 2 Release 19.0.0.0.0 - Production] Timezone: Local
=== RUN   TestForError8192
    z_test.go:4142: LMT: LMT: strconv.ParseInt: parsing "LMT": invalid syntax
--- FAIL: TestForError8192 (0.00s)
FAIL
exit status 1
FAIL	github.com/godror/godror	2.623s

```

**Reply by tgulacsi** [reply]
I've meant leaving params.Timezone as is, and using `time.Time{}.Add(-5 * time.Hour).In(time.FixedZone("LMT", +5*3600))` as the time inserted into the table.

**Reply by JesperSiteimprove** [reply]
I am a little unsure what +0050 exactly means - plus 50 minutes?

If I do
```
problemT = sql.NullTime{Time: time.Time{}.Add(-50 * time.Minute).In(time.FixedZone("LMT", (+50 * 60))), Valid: true}
_, err = stmt.Exec(sql.Named("problem_ts", problemT))
```
I get the error:
```
ersion: v0.26.3
Client: 12.2.0.1.0 Timezone: Local
Server: 19.0.0.0.0 [Oracle Database 19c Standard Edition 2 Release 19.0.0.0.0 - Production] Timezone: Local
=== RUN   TestForError8192
Time user passed 0001-01-01 00:00:00 +0050 LMT
 Timezone used in connection  Local
 Time passed to dpiData_setTimestamp  0001-01-01 00:00:20 +0050 LMT
    z_test.go:4207: exec failure: ORA-08192: Flashback-operation på tabel er ikke tilladt på faste tabeller
--- FAIL: TestForError8192 (0.06s)
FAIL
exit status 1
FAIL	github.com/godror/godror	0.221s

```

**Reply by JesperSiteimprove** [reply]
If I add 70 minutes like this:
```
problemT = sql.NullTime{Time: time.Time{}.Add(70 * time.Minute).In(time.FixedZone("LMT", (+50 * 60))), Valid: true}
```
it doesn't fail and
output is:
```
problemT:  {0001-01-01 02:00:00 +0050 LMT true}
Time user passed 0001-01-01 02:00:00 +0050 LMT
 Timezone used in connection  Local
 Time passed to dpiData_setTimestamp  0001-01-01 02:00:20 +0050 LMT
```
If I instead add 69 minutes:
```
problemT = sql.NullTime{Time: time.Time{}.Add(69 * time.Minute).In(time.FixedZone("LMT", (+50 * 60))), Valid: true}
```
it fails
and the output is:
```
problemT:  {0001-01-01 01:59:00 +0050 LMT true}
Time user passed 0001-01-01 01:59:00 +0050 LMT
 Timezone used in connection  Local
 Time passed to dpiData_setTimestamp  0001-01-01 01:59:20 +0050 LMT
    z_test.go:4208: exec failure: ORA-08192: Flashback-operation på tabel er ikke tilladt på faste tabeller
--- FAIL: TestForError8192 (0.11s)
```

**Reply by tgulacsi** [reply]
Yes, that "+0050" is 50 minutes, not 5 hours as I've read.

TL;DR; there are valid dates that Oracle seems to reject with ORA-08192.

Can we find all of them? 
Maybe a loop that decreases the time by minute and prints the wrong ones - maybe we could find some contigous region, and...
what? What should we do with those times? Inject as NULL?

**Reply by JesperSiteimprove** [reply]
I am not sure what to do anymore. I think the biggest issue is the strange "+0050" LMT timezone. 

When I set timezone=UTC on the connection parameters, then I cannot reproduce the problem.
So maybe that is just the solution?

When googling, I see others that have seen strange LMT offsets in go:
https://stackoverflow.com/questions/38853564/where-does-go-get-time-zone-information-from

**Reply by JesperSiteimprove** [reply]
this go playground exampel also shows the LMT timezone:
https://play.golang.org/p/3QqXgTwwZ1

**Comment by sudarshan12s** [comment]
Thanks for the info.   
can you also get this data.
1. tzoffset sent to function, dpiData_setTimestamp. You can enable log by adding this line
```
godror.Log = func(keyvals ...interface{}) error { fmt.Println(keyvals); return nil }
```

you will see offset in the console (ignore other prints , will be too many ) like this. below is from my mc.
[timezone -07:00 tz PST8PDT error <nil>]
[tz PST8PDT offSecs -25200]


2. I think you get are seeing this  issue even with instant client 19c. 
3. issue is seen even with column type TIMESTAMP WITH TIME ZONE?

**Reply by JesperSiteimprove** [reply]
output from the 
`godror.Log = func(keyvals ...interface{}) error { fmt.Println(keyvals); return nil }`

```
[dbTZ +01:00 dbOSTZ +02:00]
[timezone +02:00 tz Local error <nil>]
[tz Local offSecs 7200]
```

I get the same error when the column is created with time zone:
`problem_ts timestamp with time zone`

**Reply by sudarshan12s** [reply]
Thanks for sharing.  Yes offset looks wrong. 
The ORA-8192 is incorrect and we have raised an internal issue to look into correcting the error message to proper ORA error and the validation logic.

**Reply by tgulacsi** [reply]
As I see, there are two problems:
1. ORA-8192 is totally unrelated to such a "timely" timezone error
2. UTC 0000-12-31T23:59:00Z, setting Y=1 M=January D=1 h=1 m=15 s=20 with time zone tzHour=1 tzMin=16 gives an error. Why? (https://github.com/godror/godror/discussions/187#discussioncomment-1484520)

**Comment by tgulacsi** [comment]
With

```
    tim := time.Time{}.UTC() //time.FixedZone("LMT", (+50 * 60)))                                                                                              
    for i := -70 * 60; i < 70*60; i++ {                                                                                                                        
        tim := tim.Add(time.Duration(i) * time.Second)                                                                                                         
        _, err = stmt.Exec(sql.Named("problem_ts", sql.NullTime{                                                                                               
            Time:  tim,                                                                                                                                        
            Valid: true}))                                                                                                                                     
        if err != nil {                                                                                                                                        
            t.Fatalf("exec failure for %v: %v\n", tim, err)                                                                                                    
        }                                                                                                                                                      
    }          
```

I get 
    z_test.go:4201: exec failure for 0000-12-31 22:50:00 +0000 UTC: ORA-01841: (full) year must be between -4713 and +9999, and not be 0

If I treat out-of-range dates as NULL, the error goes away: 0a7f618
Does 0a7f618 help your case?

**Comment by tgulacsi** [comment]
See `SELECT TO_DATE('0001-01-01','SYYYY-MM-DD'), TO_CHAR(TO_DATE('0001-01-01','SYYYY-MM-DD')-1, 'SYYYY-MM-DD')  FROM dual`

Oracle skips year 0, 0001-01-01 follows -0001-12-31 !

"For dates before the year 1, unlike the proleptic Gregorian calendar used in the international standard ISO 8601, the traditional proleptic Gregorian calendar (like the Julian calendar) does not have a year 0 and instead uses the ordinal numbers 1, 2, ... both for years AD and BC. Thus the traditional time line is 2 BC, 1 BC, AD 1, and AD 2. ISO 8601 uses astronomical year numbering which includes a year 0 and negative numbers before it. Thus the ISO 8601 time line is −0001, 0000, 0001, and 0002. "

So I assume Oracle uses "traditional proleptic Gregorian calendar".

Can you test  6900155 ?

**Reply by JesperSiteimprove** [reply]
unfortunately it makes not difference:

```
=== RUN   TestForError8192
=== PAUSE TestForError8192
=== CONT  TestForError8192
    z_test.go:4176: problemT=0001-01-01 00:00:00 +0050 LMT, valid? true, zero? false
    z_test.go:4205: exec failure for 0001-01-01 00:00:00 +0050 LMT: ORA-08192: Flashback-operation på tabel er ikke tilladt på faste tabeller
--- FAIL: TestForError8192 (1.09s)
FAIL
exit status 1
FAIL	github.com/godror/godror	3.225s

```

**Comment by ykh-siteimprove** [comment]
I manage to reproduce the error on VM:

- Oracle Linux 7 + most recent updates
- go 1.17.2 (most recent)
- gcc version 4.8.5
- Instant Client 21.3
- Oracle Database 19.3 
- godror (most recent)


```
$ GODROR_TEST_DSN="user=sima password=*** connectString=db "  go test -v -run TestForError8192
eDSN: user=sima password=sima connectString=db
export GODROR_TEST_DSN="user=sima password=*** connectString=db configDir= connectionClass=TestClassName\nenableEvents=1 externalAuth=0 heterogeneousPool=0 libDir= newPassword= noTimezoneCheck=0\npoolIncrement=2 poolMaxSessions=16 poolMinSessions=2 poolSessionMaxLifetime=5m0s\npoolSessionTimeout=1m0s poolWaitTimeout=5s prelim=0 standaloneConnection=0\nsysasm=0 sysdba=0 sysoper=0 timezone="
godror WARNING: discrepancy between DBTIMEZONE ("+00:00"=0) and SYSTIMESTAMP ("+02:00"=200) - set connection timezone, see https://github.com/godror/godror/blob/master/doc/timezone.md
# user=sima password="SECRET-kxNufqNTUU8=" connectString=db configDir= connectionClass=TestClassName
enableEvents=1 externalAuth=0 heterogeneousPool=0 libDir= noTimezoneCheck=0
poolIncrement=2 poolMaxSessions=16 poolMinSessions=2 poolSessionMaxLifetime=5m0s
poolSessionTimeout=1m0s poolWaitTimeout=5s prelim=0 standaloneConnection=0
sysasm=0 sysdba=0 sysoper=0 timezone=
Version: v0.28.0
Client: 21.3.0.0.0 Timezone: Local
Server: 19.3.0.0.0 [Oracle Database 19c Enterprise Edition Release 19.0.0.0.0 - Production; Version 19.3.0.0.0] Timezone: Local
=== RUN   TestForError8192
=== PAUSE TestForError8192
=== CONT  TestForError8192
godror WARNING: discrepancy between DBTIMEZONE ("+00:00"=0) and SYSTIMESTAMP ("+02:00"=200) - set connection timezone, see https://github.com/godror/godror/blob/master/doc/timezone.md
    z_test.go:4176: problemT=0001-01-01 00:00:00 +0050 LMT, valid? true, zero? false
    z_test.go:4205: exec failure for 0001-01-01 00:00:00 +0050 LMT: ORA-08192: Flashback Table operation is not allowed on fixed tables
--- FAIL: TestForError8192 (0.30s)
FAIL
exit status 1
FAIL	github.com/godror/godror	0.582s
$

```

**Reply by ykh-siteimprove** [reply]
For statistics: it reproducible on Oracle Database 21c Version 21.3.0.0.0 (in premises)

**Comment by tgulacsi** [comment]
Ok, it is reproducible on our local 19.12, but still not on Oracle Autonomous Database 19.13.

    0000-12-31 22:50:00 +0000 UTC: ORA-08192: Flashback Table operation is not allowed on fixed tables

**Reply by ykh-siteimprove** [reply]
Submitted this info to our Oracle SR

**Comment by tgulacsi** [comment]
0000-12-31 22:50:00 +0000 UTC - 0001-01-01 00:43:00 +0000 UTC  
is the corrupted interval.

Strange, as it seems to go away with any FixedZone timezone (I've tried UTC, +00:50, +02:00 (my local TZ)) OR "Europe/Budapest" !
It is reproducible (locally) with time.Local as timezone ("timezone=local" in the params string)!

So the error must be there somewhere.

**Reply by tgulacsi** [reply]
```
msg=set i=0 value="sql.NullTime=sql.NullTime{Time:time.Date(0, time.December, 31, 23, 59, 0, 0, time.UTC), Valid:true}"
msg=setTimestamp time=0001-01-01T01:15:20+01:16 utc=0000-12-31T23:59:00Z tz=Local Y=1 M=January D=1 h=1 m=15 s=20 t=0 tzHour=2 tzMin=0                                                                                                                        
msg=maybeBadConn err="ORA-08192: Flashback Table operation is not allowed on fixed tables" errS="\"ORA-08192: Flashback Table operation is not allowed on fixed tables\"" errT=false errV="&godror.OraErr{message:\"Flashback Table operation is not allowed on fixed tables\", funName:\"dpiStmt_execute\", action:\"construct date\", sqlState:\"HY000\", code:8192, offset:0, recoverable:false, warning:false}"  
```

The same date with timezone=UTC:
```
msg=set i=0 value="sql.NullTime=sql.NullTime{Time:time.Date(0, time.December, 31, 23, 59, 0, 0, time.UTC), Valid:true}"
msg=setTimestamp time=0000-12-31T23:59:00Z utc=0000-12-31T23:59:00Z tz=UTC Y=-1 M=December D=31 h=23 m=59 s=0 t=0 tzHour=0 tzMin=0                                                             
```

I don't really undestand (yet) where does that "T01:15:20+01:16" coming from (it's 23:59:59Z) - it's after t.In(conn.params.Timezone()), and that timezone is Local.
BUT we call  `C.dpiConn_setTimestamp(Y=1 M=January D=1 h=1 m=15 s=20 t=0 tzHour=2 tzMin=0)` which should work (0001-01-01T15:20+02:00).

**Reply by tgulacsi** [reply]
https://play.golang.org/p/5_jUfbbJ8ej

I need to find out the why, but that LMT+0116 seems to be normal.

**Reply by tgulacsi** [reply]
https://home.kpn.nl/vanadovv/time/TZworld.html

Before 1890, Europe/Budapest did not have a time zone. For such cases the time zone database uses LMT, Local Mean Time, which calculates the time zone from the Longitude (15 degree for an hour) - for Budapest, it is 47, so +01:16; for Copenhagen, it becomes +00:50.

This is where those strange time zones coming from.


But even with the correct time zone information (not the static "2 hours"):
```
msg=set i=0 value="sql.NullTime=sql.NullTime{Time:time.Date(0, time.December, 31, 23, 59, 0, 0, time.UTC), Valid:true}"                                          
msg=setTimestamp time=0001-01-01T01:15:20+01:16 utc=0000-12-31T23:59:00Z tz=4580 Y=1 M=January D=1 h=1 m=15 s=20 t=0 tzHour=1 tzMin=16  
```

So for UTC 0000-12-31T23:59:00Z, setting Y=1 M=January D=1 h=1 m=15 s=20 with time zone tzHour=1 tzMin=16  
produces the ORA-08192 error!

**Comment by tgulacsi** [comment]
703514c6ee2e5da6078b253ddd9c1c4fc5af8f40 is a workaround.

**Reply by JesperSiteimprove** [reply]
The workaround avoids the ora-08192, put produces the wrong date in the test table:

```
select * from test_date_error;

PROBLEM_TS
-------------------
31-12-0001 23:09:40
```

The expected date was 
```
PROBLEM_TS
-------------------
01-01-0001 00:00:00
```

For now I will stick with setting the timezone=UTC in the params string.
